### ISSUE TITLE
TiKV crash with "index out of bounds" (after manual SPLIT TABLE and INSERT)

### ISSUE THREAD

**Issue by kolbe** [tikv/tikv#9774]
### What version of TiKV are you using?

```
["Welcome to TiKV"]
["Release Version:   5.0.0-rc.x"]
["Edition:           Community"]
["Git Commit Hash:   a17508c0c3a4ce62d374cf5b185b159501caa1d9"]
["Git Commit Branch: master"]
["UTC Build Time:    2021-03-08 13:25:57"]
["Rust Version:      rustc 1.51.0-nightly (1d0d76f8d 2021-01-24)"]
["Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp"]
["Profile:           dist_release"]
["memory limit in bytes: 34041469952, cpu cores quota: 8"]
```

### What operating system and CPU are you using?

AWS EKS. m5.2xlarge.

### Steps to reproduce

```
create table t1 (id int unsigned not null auto_increment , x int unsigned, primary key(id, x) clustered);
split table t1 between (0,0) and (30000,0) regions 1000;
```

```
{ printf 'insert into t1 values '; printf '(1,%d),' {1..30000} | sed 's/.$/;/'; } | mysql
```

### What did happened?

At first, things looked good. `select count(*)` returned in a short period of time:
```
MySQL [test]> select count(*) from t1;
+----------+
| count(*) |
+----------+
|    30000 |
+----------+
1 row in set (0.485 sec)
```

After checking `information_schema.tikv_region_status` a couple times to see the activity in the table's regions, I could see that a single region had all the rows, as expected:

```
MySQL [test]> select * from information_schema.tikv_region_status where approximate_size<>0 and table_name='t1' order by written_bytes desc limit 10;
+-----------+--------------------------------------------------------------------------+--------------------------------------------------------------------------+----------+---------+------------+----------+----------+------------+----------------+---------------+---------------+--
----------+------------------+------------------+-------------------------+---------------------------+
| REGION_ID | START_KEY                                                                | END_KEY                                                                  | TABLE_ID | DB_NAME | TABLE_NAME | IS_INDEX | INDEX_ID | INDEX_NAME | EPOCH_CONF_VER | EPOCH_VERSION | WRITTEN_BYTES | READ_BYTES | APPROXIMATE_SIZE | APPROXIMATE_KEYS | REPLICATIONSTATUS_STATE | REPLICATIONSTATUS_STATEID |
+-----------+--------------------------------------------------------------------------+--------------------------------------------------------------------------+----------+---------+------------+----------+----------+------------+----------------+---------------+---------------+------------+------------------+------------------+-------------------------+---------------------------+
|     17957 | 7480000000000000FF3300000000000000F8                                     | 7480000000000000FF335F720400000000FF0000001E04000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           110 |       6497853 |
  1860000 |                1 |                0 | NULL                    |                      NULL |
|     17957 | 7480000000000000FF3300000000000000F8                                     | 7480000000000000FF335F720400000000FF0000001E04000000FF0000000000000000F9 |       51 | test    | t1         |        1 |        1 | PRIMARY    |              5 |           110 |       6497853 |    1860000 |                1 |                0 | NULL                    |                      NULL |
|     19657 | 7480000000000000FF335F720400000000FF00000FF004000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF0000100E04000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           144 |          3394 |          0 |                1 |                0 | NULL                    |                      NULL |
|     18013 | 7480000000000000FF335F720400000000FF000009F604000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF00000A1404000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           144 |          3394 |          0 |                1 |                0 | NULL                    |                      NULL |
|     20813 | 7480000000000000FF335F720400000000FF000013EC04000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF0000140A04000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           161 |          3394 |
        0 |                1 |                0 | NULL                    |                      NULL |
|       105 | 7480000000000000FF335F720400000000FF000043BC04000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF000043DA04000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |            93 |          3300 |          0 |                1 |                0 | NULL                    |                      NULL |
|     17413 | 7480000000000000FF335F720400000000FF00002BD404000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF00002BF204000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           144 |          2974 |          0 |                1 |                0 | NULL                    |                      NULL |
|     13436 | 7480000000000000FF335F720400000000FF0000639C04000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF000063BA04000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           174 |          2970 |
        0 |                1 |                0 | NULL                    |                      NULL |
|     14196 | 7480000000000000FF335F720400000000FF00002DD204000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF00002DF004000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           110 |          2952 |          0 |                1 |                0 | NULL                    |                      NULL |
|     20881 | 7480000000000000FF335F720400000000FF00001DE204000000FF0000000000000000F9 | 7480000000000000FF335F720400000000FF00001E0004000000FF0000000000000000F9 |       51 | test    | t1         |        0 |     NULL | NULL       |              5 |           195 |          2782 |          0 |                1 |                0 | NULL                    |                      NULL |
+-----------+--------------------------------------------------------------------------+--------------------------------------------------------------------------+----------+---------+------------+----------+----------+------------+----------------+---------------+---------------+--
----------+------------------+------------------+-------------------------+---------------------------+
10 rows in set (0.029 sec)
```

`select count(*)` started getting really slow and eventually just returns an error:

```
MySQL [test]> select count(*) from t1;
+----------+
| count(*) |
+----------+
|    30000 |
+----------+
1 row in set (7.350 sec)

MySQL [test]> select count(*) from t1;
+----------+
| count(*) |
+----------+
|    30000 |
+----------+
1 row in set (35.001 sec)

MySQL [test]> select count(*) from t1;
ERROR 9005 (HY000): Region is unavailable
```

Meanwhile, all 3 tikv instances have crashed and cannot recover. Logs attached.

**Comment by kolbe** [comment]
Here are the logs of the initial recovery and one of the failed recovery attempts. The pods are in a crash loop.

[basic-tikv-logs.tar.gz](https://github.com/tikv/tikv/files/6111805/basic-tikv-logs.tar.gz)

**Comment by kolbe** [comment]
Initial crash node 0:
```
[2021/03/09 22:14:28.428 +00:00] [FATAL] [lib.rs:465] ["index out of bounds: the len is 2 but the index is 2"] [backtrace="stack backtrace:
   0: tikv_util::set_panic_hook::{{closure}}
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/components/tikv_util/src/lib.rs:464
   1: std::panicking::rust_panic_with_hook
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8//library/std/src/panicking.rs:595
   2: std::panicking::begin_panic_handler::{{closure}}
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8//library/std/src/panicking.rs:497
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8//library/std/src/sys_common/backtrace.rs:141
   4: rust_begin_unwind
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8//library/std/src/panicking.rs:493
   5: core::panicking::panic_fmt
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8//library/core/src/panicking.rs:92
   6: core::panicking::panic_bounds_check
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8//library/core/src/panicking.rs:69
   7: tikv::coprocessor::statistics::analyze::SampleBuilder<S>::collect_columns_stats::{{closure}}
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/src/coprocessor/statistics/analyze.rs:401
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/future/mod.rs:80
      tikv::coprocessor::statistics::analyze::AnalyzeContext<S>::handle_mixed::{{closure}}
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/src/coprocessor/statistics/analyze.rs:84
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/future/mod.rs:80
      <tikv::coprocessor::statistics::analyze::AnalyzeContext<S> as tikv::coprocessor::RequestHandler>::handle_request::__handle_request::{{closure}}
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/src/coprocessor/statistics/analyze.rs:243
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/future/mod.rs:80
  10: <core::pin::Pin<P> as core::future::future::Future>::poll
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/future/future.rs:119
      <minitrace::future::TraceWrapped<T> as core::future::future::Future>::poll
             at rust/git/checkouts/minitrace-rust-4a91d55623f07cfc/de69110/src/future.rs:107
      <tikv::coprocessor::interceptors::tracker::Tracker<F> as core::future::future::Future>::poll
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/src/coprocessor/interceptors/tracker.rs:49
  11: <tikv::coprocessor::interceptors::concurrency_limiter::ConcurrencyLimiter<PF,F> as core::future::future::Future>::poll
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/src/coprocessor/interceptors/concurrency_limiter.rs:101
  12: tikv::coprocessor::endpoint::Endpoint<E>::handle_unary_request_impl::{{closure}}
             at home/jenkins/agent/workspace/build_tikv_multi_branch_master/tikv/src/coprocessor/endpoint.rs:418
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at rustc/1d0d76f8dd4f5f6ecbeab575b87edaf1c9f56bb8/library/core/src/future/mod.rs:80
```

**Comment by BusyJay** [comment]
Fixed by #9769.
